File: progonka.m
function res = progonka()
    A = [391 -1 0 0 0 0 0 0 0 0; -4 72 -27 0 0 0 0 0 0 0; 0 7 -19 7 0 0 0 0 0 0; 0 0 -134 -460 30 0 0 0 0 0; 0 0 0 -23 281 -83 0 0 0 0;
    0 0 0 0 -374 896 -353 0 0 0; 0 0 0 0 0 -238 -716 86 0 0; 0 0 0 0 0 0 -66 147 63 0; 0 0 0 0 0 0 0 109 458 -200; 0 0 0 0 0 0 0 0 626 721];
    B = [969; 718; 571; 27; -645; -203; -733; -939; 879; -398];
    [a, b] = size(A);
    b = get_b_array(A);
    c = get_c_array(A);
    d = get_d_array(A);
    tau = 0;
    sigma = 0;
    sigmas = zeros(1, a - 1);
    lamda = 0;
    lamdas = zeros(1, a);
    i = 1;
    while i <= a
        if i == 1
            tau = get_tau(0, 0, c(i));
        else
  
            tau = get_tau(b(i - 1), sigma, c(i));
        end
        
        if i ~= a 
            sigma = get_sigma(d(i), tau);
            sigmas(i) = sigma;
        end
        
        if i == 1
            lamda = get_lamda(B(i), 0, 0, tau);
        else
            lamda = get_lamda(B(i), b(i - 1), lamda, tau);
        end
        lamdas(i) = lamda;
        i = i + 1;
    end
    answers = zeros(1, a);
    i = a;
    while i > 0
        if i == a
            x = lamdas(i);
        else
            x = sigmas(i)*answers(i+1) + lamdas(i);
        end
        answers(i) = x;
        i = i - 1;
    end
    
    res = answers;
end

function res = get_d_array(matrix)
    [a b] = size(matrix);    
    res = zeros(1, a - 1);
    i = 1;
    while i < a
        res(i) = matrix(i, i + 1);
        i = i +1;
    end
end

function res = get_c_array(matrix)
    [a b] = size(matrix);    
    res = zeros(1, a);
    i = 1;
    while i <= a
        res(i) = matrix(i, i);
        i = i + 1;
    end
end

function res = get_b_array(matrix)
    [a b] = size(matrix);    
    res = zeros(1, a - 1);
    i = 1;
    while i < a
        res(i) = matrix(i + 1, i);
        i = i +1;
    end
end

function res = get_tau(b, sigma, c)
    res = b*sigma + c;
end

function res = get_sigma(d, tau)
    res = -d/tau;
end

function res = get_lamda(r, b, lamda, tau)
    res = (r - lamda*b)/tau;
end

File: progonka.py
import datetime
from matrixoperation import show_matrix
from matrix import *


# def show_matrix(m):
#     for line in m:
#         result_line = ''
#         for elem in line:
#             result = ''
#             current = format(elem, '.2f')
#             result += '    ' + current if current.startswith('-') else '     ' + current
#             result_line += result
#         print result_line


def get_d_array(matrix):
    result = []
    i = 0
    while i < len(matrix) - 1:
        result.append(matrix[i][i + 1])
        i += 1
    return result


def get_c_array(matrix):
    result = []
    i = 0
    while i < len(matrix):
        result.append(matrix[i][i])
        i += 1
    return result


def get_b_array(matrix):
    result = []
    i = 0
    while i < len(matrix) - 1:
        result.append(matrix[i + 1][i])
        i += 1
    return result


def get_tau(b, sigma, c):
    return b*sigma + c


def get_sigma(d, tau):
    return -d/tau


def get_lamda(r, b, lamda, tau):
    return (r - lamda*b)/tau


def progonka(m, bbb):
    b = get_b_array(m)
    c = get_c_array(m)
    d = get_d_array(m)
    tau = 0.0
    sigma = 0.0
    lamda = 0.0
    lamdas = []
    sigmas = []
    i = 0
    while i < len(m):
        print 'Iteration #' + str(i+1)
        if i == 0:
            tau = get_tau(0, 0, c[i]*1.0)
        else:
            tau = get_tau(b[i - 1]*1.0, sigma*1.0, c[i]*1.0)
        print 'Tau: ' + str(tau)

        if i != len(m) - 1:
            sigma = get_sigma(d[i]*1.0, tau*1.0)
            # sigmas.append(sigma)
        print 'Sigma: ' + str(sigma)
        sigmas.append(sigma)

        if i == 0:
            lamda = get_lamda(bbb[i]*1.0, 0, 0, tau*1.0)
        else:
            lamda = get_lamda(bbb[i]*1.0, b[i - 1]*1.0, lamda*1.0, tau*1.0)
        print 'Lamda: ' + str(lamda)
        lamdas.append(lamda)
        i += 1
        print '=============================='
        print '=============================='
    # print lamdas

    answers = []
    i = len(lamdas) - 1
    while i >= 0:
        if i == len(lamdas) - 1:
            x = lamdas[i]
        else:
            x = sigmas[i]*1.0*answers[len(answers) - 1] + lamdas[i]*1.0
        answers.append(x)
        i -= 1

    for ans in answers:
        print ans


m = get_matrix3()
b = get_b3()
start = datetime.datetime.now()
progonka(m, b)
print datetime.datetime.now() - start
